CLAUDE.md

```diff
@@ -0,0 +1,134 @@
+# CLAUDE.md
+
+This file provides guidance to Claude Code (claude.ai/code) when working with code in this repository.
+
+## Project Overview
+
+Lance is a modern columnar data format optimized for ML workflows and datasets. It provides:
+- 100x faster random access than Parquet
+- Native vector search capabilities (IVF, HNSW, PQ)
+- Zero-copy automatic versioning
+- Deep integration with Apache Arrow, Pandas, Polars, DuckDB, Ray
+- Cloud-native with S3, Azure, GCS support
+
+## Project Vision
+
+The de facto standard columnar data format for machine learning and large language models.
+
+## Project Requirements
+
+- Always use English in code, examples, and comments.
+- Features should be implemented concisely, maintainably, and efficiently.
+- Code is not just for execution, but also for readability.
+- Only add meaingful comments and tests.
+
+## Architecture
+
+The project is organized as a Rust workspace with Python and Java bindings:
+
+- `rust/lance/` - Main Lance library implementing the columnar format
+- `rust/lance-arrow/` - Apache Arrow integration layer
+- `rust/lance-core/` - Core types, traits, and utilities
+- `rust/lance-encoding/` - Data encoding and compression algorithms
+- `rust/lance-file/` - File format reading/writing
+- `rust/lance-index/` - Vector and scalar indexing implementations
+- `rust/lance-io/` - I/O operations and object store integration
+- `rust/lance-linalg/` - Linear algebra operations for vector search
+- `rust/lance-table/` - Table format and operations
+- `rust/lance-datafusion/` - DataFusion query engine integration
+- `python/` - Python bindings using PyO3/maturin
+- `java/` - Java bindings using JNI
+
+## Common Development Commands
+
+### Rust Development
+
+```bash
+# Building
+cargo build                          # Debug build
+cargo build --release               # Release build
+cargo build --features <features>   # Build with specific features
+
+# Testing
+cargo test                          # Run all tests
+cargo test <test_name>             # Run specific test
+cargo test --package <package>     # Test specific crate
+cargo test -- --nocapture          # Show println! output
+
+# Formatting & Linting
+cargo fmt                          # Format code
+cargo fmt -- --check              # Check formatting
+cargo clippy -- -D warnings       # Run linter
+
+# Dependency Checking
+cargo deny check                   # Audit dependencies
+```
```

Comment:
To save on context window, I would condense this:

* Check for build errors: `cargo check --all --tests --benches`
* Run tests: `cargo test`
* Run specific test: `cargo test -p <package> <test_name>`
* Lint: `cargo clippy --all --tests --benches -- -D warnings`
* Format: `cargo fmt --all`